Framework/Algorithms/src/GetDetectorOffsets.cpp

```diff
@@ -124,59 +142,79 @@ void GetDetectorOffsets::exec() {
   Progress prog(this, 0.0, 1.0, nspec);
-  auto &spectrumInfo = maskWS->mutableSpectrumInfo();
   PARALLEL_FOR_IF(Kernel::threadSafe(*inputW))
   for (int64_t wi = 0; wi < nspec; ++wi) {
     PARALLEL_START_INTERRUPT_REGION
+    // Get the list of detectors in this pixel
+    const auto &dets = inputW->getSpectrum(wi).getDetectorIDs();
+
+    // If the entire spectrum is already masked, there's nothing to do.
+    if (maskWS->isMasked(dets))
+      continue;
+
     // Fit the peak
     double offset = fitSpectra(wi);
-    double mask = 0.0;
+    bool spectrumIsMasked = false;
     if (std::abs(offset) > m_maxOffset) {
-      offset = 0.0;
-      mask = 1.0;
+      g_log.debug() << "[GetDetectorOffsets]: fit failure: offset: " << std::abs(offset)
+                    << " is greater than maximum allowed: " << m_maxOffset << "." << std::endl;
+      spectrumIsMasked = true;
     }
 
-    // Get the list of detectors in this pixel
-    const auto &dets = inputW->getSpectrum(wi).getDetectorIDs();
-
     // Most of the exec time is in FitSpectra, so this critical block should not
     // be a problem.
     PARALLEL_CRITICAL(GetDetectorOffsets_setValue) {
       // Use the same offset for all detectors from this pixel
       for (const auto &det : dets) {
-        outputW->setValue(det, offset);
-        const auto mapEntry = pixel_to_wi.find(det);
-        if (mapEntry == pixel_to_wi.end())
+        if (spectrumIsMasked) {
+          maskWS->setMasked(det, true);
           continue;
-        const size_t workspaceIndex = mapEntry->second;
-        if (mask == 1.) {
-          // Being masked
-          maskWS->getSpectrum(workspaceIndex).clearData();
-          spectrumInfo.setMasked(workspaceIndex, true);
-          maskWS->mutableY(workspaceIndex)[0] = mask;
-        } else {
-          // Using the detector
-          maskWS->mutableY(workspaceIndex)[0] = mask;
         }
+        // Warning: individual detectors in a spectrum may be masked.
+        if (!maskWS->isMasked(det))
+          outputW->setValue(det, offset);
       }
     }
     prog.report();
     PARALLEL_END_INTERRUPT_REGION
   }
   PARALLEL_CHECK_INTERRUPT_REGION
-
+  // Make sure that the output workspaces' detector masks are consistent with the mask values.
+  maskWS->combineToDetectorMasks();
+  maskWS->combineToDetectorMasks(outputW->mutableDetectorInfo());
+
+  if (g_log.getLevel() >= Logger::Priority::PRIO_TRACE) {
+    auto &trace(g_log.getLogStream(Logger::Priority::PRIO_TRACE));
+    trace << "[GetDetectorOffsets]: Computed offsets:" << std::endl;
+    for (size_t ns = 0; ns < inputW->getNumberHistograms(); ++ns) {
+      const auto dets = inputW->getSpectrum(ns).getDetectorIDs();
+      for (const auto &det : dets)
+        trace << "  " << outputW->getValue(det) << (maskWS->isMasked(det) ? "*" : "") << std::endl;
+    }
+  }
   // Return the output
   setProperty("OutputWorkspace", outputW);
-  setProperty("MaskWorkspace", maskWS);
```

Comment:
I think this `setProperty()` still needs to be there to update the pointer/reference in case it was changed above.